File: parser/src/main/java/com/github/nsc/de/shake/generation/ShakeGenerator.java
package com.github.nsc.de.shake.generation;

import com.github.nsc.de.shake.parser.node.*;
import com.github.nsc.de.shake.parser.node.expression.*;
import com.github.nsc.de.shake.parser.node.factor.DoubleNode;
import com.github.nsc.de.shake.parser.node.factor.IntegerNode;
import com.github.nsc.de.shake.parser.node.functions.FunctionCallNode;
import com.github.nsc.de.shake.parser.node.functions.FunctionDeclarationNode;
import com.github.nsc.de.shake.parser.node.logical.*;
import com.github.nsc.de.shake.parser.node.loops.DoWhileNode;
import com.github.nsc.de.shake.parser.node.loops.ForNode;
import com.github.nsc.de.shake.parser.node.loops.WhileNode;
import com.github.nsc.de.shake.parser.node.objects.ClassConstructionNode;
import com.github.nsc.de.shake.parser.node.objects.ClassDeclarationNode;
import com.github.nsc.de.shake.parser.node.variables.*;

public abstract class ShakeGenerator<T> implements ShakeGeneratorBase {
    
    public T visit(Node n) {

        if(n instanceof Tree) return visitTree((Tree) n);
        if(n instanceof DoubleNode) return visitDoubleNode((DoubleNode) n);
        if(n instanceof IntegerNode) return visitIntegerNode((IntegerNode) n);
        if(n instanceof AddNode) return visitAddNode((AddNode) n);
        if(n instanceof SubNode) return visitSubNode((SubNode) n);
        if(n instanceof MulNode) return visitMulNode((MulNode) n);
        if(n instanceof DivNode) return visitDivNode((DivNode) n);
        if(n instanceof ModNode) return visitModNode((ModNode) n);
        if(n instanceof PowNode) return visitPowNode((PowNode) n);
        if(n instanceof VariableDeclarationNode) return visitVariableDeclarationNode((VariableDeclarationNode) n);
        if(n instanceof VariableAddAssignmentNode) return visitVariableAddAssignmentNode((VariableAddAssignmentNode) n);
        if(n instanceof VariableSubAssignmentNode) return visitVariableSubAssignmentNode((VariableSubAssignmentNode) n);
        if(n instanceof VariableMulAssignmentNode) return visitVariableMulAssignmentNode((VariableMulAssignmentNode) n);
        if(n instanceof VariableDivAssignmentNode) return visitVariableDivAssignmentNode((VariableDivAssignmentNode) n);
        if(n instanceof VariableModAssignmentNode) return visitVariableModAssignmentNode((VariableModAssignmentNode) n);
        if(n instanceof VariablePowAssignmentNode) return visitVariablePowAssignmentNode((VariablePowAssignmentNode) n);
        if(n instanceof VariableIncreaseNode) return visitVariableIncreaseNode((VariableIncreaseNode) n);
        if(n instanceof VariableDecreaseNode) return visitVariableDecreaseNode((VariableDecreaseNode) n);
        if(n instanceof VariableAssignmentNode) return visitVariableAssignmentNode((VariableAssignmentNode) n);
        if(n instanceof VariableUsageNode) return visitVariableUsageNode((VariableUsageNode) n);
        if(n instanceof LogicalEqEqualsNode) return visitEqEqualsNode((LogicalEqEqualsNode) n);
        if(n instanceof LogicalBiggerEqualsNode) return visitBiggerEqualsNode((LogicalBiggerEqualsNode) n);
        if(n instanceof LogicalSmallerEqualsNode) return visitSmallerEqualsNode((LogicalSmallerEqualsNode) n);
        if(n instanceof LogicalBiggerNode) return visitBiggerNode((LogicalBiggerNode) n);
        if(n instanceof LogicalSmallerNode) return visitSmallerNode((LogicalSmallerNode) n);
        if(n instanceof LogicalAndNode) return visitLogicalAndNode((LogicalAndNode) n);
        if(n instanceof LogicalOrNode) return visitLogicalOrNode((LogicalOrNode) n);
        if(n instanceof LogicalXOrNode) return visitLogicalXOrNode((LogicalXOrNode) n);
        if(n instanceof WhileNode) return visitWhileNode((WhileNode) n);
        if(n instanceof DoWhileNode) return visitDoWhileNode((DoWhileNode) n);
        if(n instanceof ForNode) return visitForNode((ForNode) n);
        if(n instanceof IfNode) return visitIfNode((IfNode) n);
        if(n instanceof FunctionDeclarationNode) return visitFunctionDeclarationNode((FunctionDeclarationNode) n);
        if(n instanceof ClassConstructionNode) return visitClassConstruction((ClassConstructionNode) n);
        if(n instanceof FunctionCallNode) return visitFunctionCallNode((FunctionCallNode) n);
        if(n instanceof IdentifierNode) return visitIdentifierNode((IdentifierNode) n);
        if(n instanceof ClassDeclarationNode) return visitClassDeclarationNode((ClassDeclarationNode) n);
        if(n instanceof LogicalTrueNode) return visitLogicalTrueNode((LogicalTrueNode) n);
        if(n instanceof LogicalFalseNode) return visitLogicalFalseNode((LogicalFalseNode) n);
        if(n instanceof ImportNode) return visitImportNode((ImportNode) n);
        if(n instanceof CastNode) return visitCastNode((CastNode) n);
        throw new Error(String.format("It looks like that node is not implemented in the Interpreter: %s", n.getClass().toString()));

    }

    public abstract T visitTree(Tree t);
    public abstract T visitDoubleNode(DoubleNode n);
    public abstract T visitIntegerNode(IntegerNode n);
    public abstract T visitAddNode(AddNode n);
    public abstract T visitSubNode(SubNode n);
    public abstract T visitMulNode(MulNode n);
    public abstract T visitDivNode(DivNode n);
    public abstract T visitModNode(ModNode n);
    public abstract T visitPowNode(PowNode n);
    public abstract T visitVariableDeclarationNode(VariableDeclarationNode n);
    public abstract T visitVariableAssignmentNode(VariableAssignmentNode n);
    public abstract T visitVariableAddAssignmentNode(VariableAddAssignmentNode n);
    public abstract T visitVariableSubAssignmentNode(VariableSubAssignmentNode n);
    public abstract T visitVariableMulAssignmentNode(VariableMulAssignmentNode n);
    public abstract T visitVariableDivAssignmentNode(VariableDivAssignmentNode n);
    public abstract T visitVariableModAssignmentNode(VariableModAssignmentNode n);
    public abstract T visitVariablePowAssignmentNode(VariablePowAssignmentNode n);
    public abstract T visitVariableIncreaseNode(VariableIncreaseNode n);
    public abstract T visitVariableDecreaseNode(VariableDecreaseNode n);
    public abstract T visitVariableUsageNode(VariableUsageNode n);
    public abstract T visitEqEqualsNode(LogicalEqEqualsNode n);
    public abstract T visitBiggerEqualsNode(LogicalBiggerEqualsNode n);
    public abstract T visitSmallerEqualsNode(LogicalSmallerEqualsNode n);
    public abstract T visitBiggerNode(LogicalBiggerNode n);
    public abstract T visitSmallerNode(LogicalSmallerNode n);
    public abstract T visitLogicalAndNode(LogicalAndNode n);
    public abstract T visitLogicalOrNode(LogicalOrNode n);
    public abstract T visitLogicalXOrNode(LogicalXOrNode n);
    public abstract T visitWhileNode(WhileNode n);
    public abstract T visitDoWhileNode(DoWhileNode n);
    public abstract T visitForNode(ForNode n);
    public abstract T visitIfNode(IfNode n);
    public abstract T visitFunctionDeclarationNode(FunctionDeclarationNode n);
    public abstract T visitClassDeclarationNode(ClassDeclarationNode n);
    public abstract T visitClassConstruction(ClassConstructionNode n);
    public abstract T visitFunctionCallNode(FunctionCallNode n);
    public abstract T visitIdentifierNode(IdentifierNode n);
    public abstract T visitLogicalTrueNode(LogicalTrueNode n);
    public abstract T visitLogicalFalseNode(LogicalFalseNode n);
    public abstract T visitImportNode(ImportNode n);
    public abstract T visitCastNode(CastNode n);

}

File: parser/src/commonMain/kotlin/com/github/nsc/de/shake/generation/ShakeGenerator.kt
package com.github.nsc.de.shake.generation

import com.github.nsc.de.shake.parser.node.*
import com.github.nsc.de.shake.parser.node.factor.IntegerNode
import com.github.nsc.de.shake.parser.node.variables.VariableDeclarationNode
import com.github.nsc.de.shake.parser.node.variables.VariableAddAssignmentNode
import com.github.nsc.de.shake.parser.node.variables.VariableSubAssignmentNode
import com.github.nsc.de.shake.parser.node.variables.VariableMulAssignmentNode
import com.github.nsc.de.shake.parser.node.variables.VariableDivAssignmentNode
import com.github.nsc.de.shake.parser.node.variables.VariableModAssignmentNode
import com.github.nsc.de.shake.parser.node.variables.VariablePowAssignmentNode
import com.github.nsc.de.shake.parser.node.variables.VariableIncreaseNode
import com.github.nsc.de.shake.parser.node.variables.VariableDecreaseNode
import com.github.nsc.de.shake.parser.node.variables.VariableAssignmentNode
import com.github.nsc.de.shake.parser.node.variables.VariableUsageNode
import com.github.nsc.de.shake.parser.node.logical.LogicalEqEqualsNode
import com.github.nsc.de.shake.parser.node.logical.LogicalBiggerEqualsNode
import com.github.nsc.de.shake.parser.node.logical.LogicalSmallerEqualsNode
import com.github.nsc.de.shake.parser.node.logical.LogicalBiggerNode
import com.github.nsc.de.shake.parser.node.logical.LogicalSmallerNode
import com.github.nsc.de.shake.parser.node.logical.LogicalAndNode
import com.github.nsc.de.shake.parser.node.logical.LogicalOrNode
import com.github.nsc.de.shake.parser.node.logical.LogicalXOrNode
import com.github.nsc.de.shake.parser.node.loops.WhileNode
import com.github.nsc.de.shake.parser.node.loops.DoWhileNode
import com.github.nsc.de.shake.parser.node.loops.ForNode
import com.github.nsc.de.shake.parser.node.functions.FunctionDeclarationNode
import com.github.nsc.de.shake.parser.node.objects.ClassConstructionNode
import com.github.nsc.de.shake.parser.node.functions.FunctionCallNode
import com.github.nsc.de.shake.parser.node.objects.ClassDeclarationNode
import com.github.nsc.de.shake.parser.node.logical.LogicalTrueNode
import com.github.nsc.de.shake.parser.node.logical.LogicalFalseNode
import com.github.nsc.de.shake.parser.node.expression.*
import com.github.nsc.de.shake.parser.node.factor.DoubleNode

abstract class ShakeGenerator<T> : ShakeGeneratorBase {
    fun visit(n: Node): T {
        if (n is Tree) return visitTree(n)
        if (n is DoubleNode) return visitDoubleNode(n)
        if (n is IntegerNode) return visitIntegerNode(n)
        if (n is AddNode) return visitAddNode(n)
        if (n is SubNode) return visitSubNode(n)
        if (n is MulNode) return visitMulNode(n)
        if (n is DivNode) return visitDivNode(n)
        if (n is ModNode) return visitModNode(n)
        if (n is PowNode) return visitPowNode(n)
        if (n is VariableDeclarationNode) return visitVariableDeclarationNode(n)
        if (n is VariableAddAssignmentNode) return visitVariableAddAssignmentNode(n)
        if (n is VariableSubAssignmentNode) return visitVariableSubAssignmentNode(n)
        if (n is VariableMulAssignmentNode) return visitVariableMulAssignmentNode(n)
        if (n is VariableDivAssignmentNode) return visitVariableDivAssignmentNode(n)
        if (n is VariableModAssignmentNode) return visitVariableModAssignmentNode(n)
        if (n is VariablePowAssignmentNode) return visitVariablePowAssignmentNode(n)
        if (n is VariableIncreaseNode) return visitVariableIncreaseNode(n)
        if (n is VariableDecreaseNode) return visitVariableDecreaseNode(n)
        if (n is VariableAssignmentNode) return visitVariableAssignmentNode(n)
        if (n is VariableUsageNode) return visitVariableUsageNode(n)
        if (n is LogicalEqEqualsNode) return visitEqEqualsNode(n)
        if (n is LogicalBiggerEqualsNode) return visitBiggerEqualsNode(n)
        if (n is LogicalSmallerEqualsNode) return visitSmallerEqualsNode(n)
        if (n is LogicalBiggerNode) return visitBiggerNode(n)
        if (n is LogicalSmallerNode) return visitSmallerNode(n)
        if (n is LogicalAndNode) return visitLogicalAndNode(n)
        if (n is LogicalOrNode) return visitLogicalOrNode(n)
        if (n is LogicalXOrNode) return visitLogicalXOrNode(n)
        if (n is WhileNode) return visitWhileNode(n)
        if (n is DoWhileNode) return visitDoWhileNode(n)
        if (n is ForNode) return visitForNode(n)
        if (n is IfNode) return visitIfNode(n)
        if (n is FunctionDeclarationNode) return visitFunctionDeclarationNode(n)
        if (n is ClassConstructionNode) return visitClassConstruction(n)
        if (n is FunctionCallNode) return visitFunctionCallNode(n)
        if (n is IdentifierNode) return visitIdentifierNode(n)
        if (n is ClassDeclarationNode) return visitClassDeclarationNode(n)
        if (n is LogicalTrueNode) return visitLogicalTrueNode(n)
        if (n is LogicalFalseNode) return visitLogicalFalseNode(n)
        if (n is ImportNode) return visitImportNode(n)
        if (n is CastNode) return visitCastNode(n)
        throw Error("It looks like that node is not implemented in the Interpreter: $n")
    }

    abstract fun visitTree(t: Tree?): T
    abstract fun visitDoubleNode(n: DoubleNode?): T
    abstract fun visitIntegerNode(n: IntegerNode?): T
    abstract fun visitAddNode(n: AddNode?): T
    abstract fun visitSubNode(n: SubNode?): T
    abstract fun visitMulNode(n: MulNode?): T
    abstract fun visitDivNode(n: DivNode?): T
    abstract fun visitModNode(n: ModNode?): T
    abstract fun visitPowNode(n: PowNode?): T
    abstract fun visitVariableDeclarationNode(n: VariableDeclarationNode?): T
    abstract fun visitVariableAssignmentNode(n: VariableAssignmentNode?): T
    abstract fun visitVariableAddAssignmentNode(n: VariableAddAssignmentNode?): T
    abstract fun visitVariableSubAssignmentNode(n: VariableSubAssignmentNode?): T
    abstract fun visitVariableMulAssignmentNode(n: VariableMulAssignmentNode?): T
    abstract fun visitVariableDivAssignmentNode(n: VariableDivAssignmentNode?): T
    abstract fun visitVariableModAssignmentNode(n: VariableModAssignmentNode?): T
    abstract fun visitVariablePowAssignmentNode(n: VariablePowAssignmentNode?): T
    abstract fun visitVariableIncreaseNode(n: VariableIncreaseNode?): T
    abstract fun visitVariableDecreaseNode(n: VariableDecreaseNode?): T
    abstract fun visitVariableUsageNode(n: VariableUsageNode?): T
    abstract fun visitEqEqualsNode(n: LogicalEqEqualsNode?): T
    abstract fun visitBiggerEqualsNode(n: LogicalBiggerEqualsNode?): T
    abstract fun visitSmallerEqualsNode(n: LogicalSmallerEqualsNode?): T
    abstract fun visitBiggerNode(n: LogicalBiggerNode?): T
    abstract fun visitSmallerNode(n: LogicalSmallerNode?): T
    abstract fun visitLogicalAndNode(n: LogicalAndNode?): T
    abstract fun visitLogicalOrNode(n: LogicalOrNode?): T
    abstract fun visitLogicalXOrNode(n: LogicalXOrNode?): T
    abstract fun visitWhileNode(n: WhileNode?): T
    abstract fun visitDoWhileNode(n: DoWhileNode?): T
    abstract fun visitForNode(n: ForNode?): T
    abstract fun visitIfNode(n: IfNode?): T
    abstract fun visitFunctionDeclarationNode(n: FunctionDeclarationNode?): T
    abstract fun visitClassDeclarationNode(n: ClassDeclarationNode?): T
    abstract fun visitClassConstruction(n: ClassConstructionNode?): T
    abstract fun visitFunctionCallNode(n: FunctionCallNode?): T
    abstract fun visitIdentifierNode(n: IdentifierNode?): T
    abstract fun visitLogicalTrueNode(n: LogicalTrueNode?): T
    abstract fun visitLogicalFalseNode(n: LogicalFalseNode?): T
    abstract fun visitImportNode(n: ImportNode?): T
    abstract fun visitCastNode(n: CastNode?): T
}